### PR TITLE
Faster lookup

### PR DESCRIPTION
**Pull request by privat**
This series improves the lookup strategies:

* special fast track for the virtual type SELF
* choose to iterate on the mclassdefs instead of the mpropdefs if they are less numerous.

This mainly limit the degenerative cases that where discovered while investigating the slowdown of #2223

with nitc/nitc/nitc:
before: 0m7.168s
after: 0m6.232s

with nitpick ../contrib:
before: 0m20.928s
after: 0m19.432s

### CODE REVIEW

**Comment by R4PaSs** [comment]
:shipit:

File: src/model/model.nit [hunk 1]
```diff
@@ -722,6 +722,11 @@ class MClassDef
 
 	# All property definitions in the class (introductions and redefinitions)
 	var mpropdefs = new Array[MPropDef]
+
+	# All property definitions in the class by its associated property.
+	#
+	# This include introductions and redefinitions, not inheritance.
+	var mpropdefs_by_property = new HashMap[MProperty, MPropDef]
```
Comment (jcbrinfo) [inline src/model/model.nit]: Not sure for the documentation. For what I see, this dictionary only includes `MPropDef`s of the class definition, not of the entire class.

Comment (privat) [inline src/model/model.nit]: right, I copy-pasted the doc of the previous property, that was also missing the `definition` part. maybe replace `the class` by `self`.

**Comment by privat** [comment]
reroll: try to improve the documentation of `MClassDef::mproprdefs` and `MClassDef::mproprdefs_by_property`

**Comment by privat** [comment]
test bootstrap please

**Comment by jcbrinfo** [comment]
:shipit: :racehorse:

**Comment by privat** [comment]
test bootstrap please

**Pull request merged**